### PR TITLE
Conflicts: Add user name to conflict file name

### PR DESCRIPTION
**Pull request by ckamm**
For the case of uploading conflict files only.

See #6325. Possible issue is too-long-path on windows.

### CODE REVIEW

**Comment by ogoffart** [comment]
Btw, Jan has been insisting in the past that the filename should not contains these wierd '_' and '-' that nobody understand.   Instead, it should maybe look like this more user friendly `" (Conflict UserName 2018-02-17 14.53.20)"`  name

Compare:

```
Readme  (Conflict OlivierGoffart 2018-02-17 14.53.20).txt
Readme_conflict-OlivierGoffart-20180217145320.txt
```

I guess now is time to do this change.

File: src/common/utility.cpp [hunk 1]
```diff
@@ -595,4 +602,12 @@ QByteArray Utility::conflictFileBaseName(const QByteArray &conflictName)
     return "";
 }
 
+QString Utility::sanitizeForFileName(const QString &name)
+{
+    static QRegularExpression regex(QStringLiteral("[\x00-\x1f\x80-\x9f/?<>\\\\:*|\"]"));
```
Comment (ogoffart) [inline src/common/utility.cpp]:
```
QString invalid("?<>\\:*|...") 
... std::remove_if(..., [&](auto c) { return !c.isPrint() || invalid.contains(c); }  ) ...
```

File: test/testutility.cpp [hunk 2]
```diff
@@ -187,7 +187,28 @@ private slots:
         qunsetenv("OWNCLOUD_TEST_CASE_PRESERVING");
     }
 
+    void testSanitizeForFileName_data()
+    {
+        QTest::addColumn<QString>("input");
+        QTest::addColumn<QString>("output");
+
+        QTest::newRow("")
```
Comment (ogoffart) [inline test/testutility.cpp]: minor: maybe give title to the data

**Comment by ckamm** [comment]
The more user-friendly conflict tags are also what dropbox does. Let's make a different ticket about changing our tags though.

File: src/common/utility.cpp (hunk 1 above)
Comment (ckamm) [reply src/common/utility.cpp]: That would also drop compositing unicode codepoints as far as I can tell.

File: src/common/utility.cpp [hunk 3]
```diff
@@ -36,7 +36,7 @@
 #include <QStandardPaths>
 #include <QCollator>
 #include <QSysInfo>
-
+#include <QRegularExpression>
```
Comment (ogoffart) [inline src/common/utility.cpp]: (leftover?)

File: src/common/utility.cpp [hunk 4]
```diff
@@ -595,4 +602,19 @@ QByteArray Utility::conflictFileBaseName(const QByteArray &conflictName)
     return "";
 }
 
+QString Utility::sanitizeForFileName(const QString &name)
+{
+    const auto invalid = QStringLiteral("/?<>\\:*|\"");
+    QString result;
+    result.reserve(name.size());
+    for (const auto c : name) {
```
Comment (ogoffart) [inline src/common/utility.cpp]: you did not like the use of std::remove_if?

Comment (ckamm) [reply src/common/utility.cpp]: Yes, I didn't like it at all. I thought `remove_if` followed by `truncate(distance(begin, newend))` looked way worse than this

**Comment by ogoffart** [comment]
Ah, this seems to have broken the tests:

15: FAIL!  : TestSyncConflict::testUploadAfterDownload() 'conflictMap[a1FileId].contains(QString("-%1-").arg(fakeFolder.syncEngine().account()->davUser()))' returned FALSE. ()

**Comment by ckamm** [comment]
Noooo. Will fix now.

**Pull request merged**